### PR TITLE
Try to improve AI::MoveTo() and AI::StoppingPoint() with reverse thrusters

### PR DESCRIPTION
**Pull request by wjp**
**Feature:** This PR implements the feature request detailed and discussed in issue #7577

## Feature Details

This patch adds drag calculations to StoppingPoint to make that more accurate. StoppingPoint now also returns the direction to accelerate in to reach the desired velocity and the time it takes to reach it. It removes the choice of forward/reverse thrusters from StoppingPoint and moves that responsibility to MoveTo to simplify the code.
    
The new StoppingPoint is used in a changed AI::MoveTo function that now tries to more robustly select the type of approach used. Roughly speaking the choices are forward-then-flip, forward-then-reverse, full-reverse, with a special case for reverse-then-forward for landing immediately after launch.
    
Notably having full-reverse plans means the auto-pilot can now reliably land ships that only have reverse thrusters.
    
Finally, MoveTo now doesn't send zero degree turn commands anymore to prevent ships with visible steering thrusters effects from triggering those continuously.
    
None of these changes lead to "perfect" executions of approaches; this code still very much uses the "on the fly" method where the approach gets corrected while getting nearer to the target.

The commits in this branch also add some debugging code that draws arrows and labels for the planned approach for any ship with the name "debug". The last commit (before the style fixes :-) ) in the branch removes all traces of it so squashing everything together should leave no debugging code.

## UI Screenshots
N/A

## Usage Examples
N/A

## Testing Done
Extensive landing on planets using various ships from various angles and relative speeds. This savegame has a small selection of ships with reverse thrusters available (borrowed from other issues and my own playthroughs): [debug approach.txt](https://github.com/endless-sky/endless-sky/files/10087110/debug.approach.txt)

### Automated Tests Added
I have not added any tests.

## Performance Impact
While the new `StoppingPoint` function is more complicated than the old one, the added math is limited to a sqrt and a log (and slightly more multiplications and divisions) so the impact should be limited.

### CODE REVIEW

**Comment by wjp** [comment]
https://user-images.githubusercontent.com/327756/203862546-edd84b28-e1d6-469d-8162-81edc44ea846.mp4

**Comment by wjp** [comment]
Similar landings using current master:

https://user-images.githubusercontent.com/327756/203863630-d40a719f-ed36-438f-8520-fd49da39491f.mp4

File: source/AI.cpp [hunk 1]
```diff
@@ -2810,56 +2948,105 @@ bool AI::DoSecretive(Ship &ship, Command &command)
...
 	Point position = ship.Position();
-	Point velocity = ship.Velocity() - targetVelocity;
+	Point velocity = ship.Velocity();
 	Angle angle = ship.Facing();
-	double acceleration = ship.Acceleration();
+	double acceleration;
+	if(!reverse)
+		acceleration = ship.Acceleration();
+	else
+		acceleration = ship.Attributes().Get("reverse thrust") / ship.InertialMass();
```
Comment (warp-core) [inline source/AI.cpp:2960]: Maybe add a `ReverseAcceleration()` method to `Ship`?

File: source/AI.cpp [hunk 2]
```diff
@@ -2810,56 +2948,105 @@ bool AI::DoSecretive(Ship &ship, Command &command)
...
+
 	double turnRate = ship.TurnRate();
-	shouldReverse = false;
 
 	// If I were to turn around and stop now the relative movement, where would that put me?
-	double v = velocity.Length();
-	if(!v)
+
+	double v = (velocity - targetVelocity).Length();
+	if(v < 1e-6)
```
Comment (warp-core) [inline source/AI.cpp:2967]: You use this `1e-6` a few times, maybe put it in a `static const` variable at the top of this method?
Also, I think it should be `1.e-6`, for consistency?

**Comment by Amazinite** [comment]
@wjp Are you able to get to Warp's review and resolve the conflicts of this PR?

**Comment by Amazinite** [comment]
Marking both #7703 and #8354 as orphaned and closing them. There's a lot of agreement that the reverse thruster AI needs improved upon, and both of these PRs do make improvements upon it, but I think they each miss the mark in different ways. Therefore, a combination of the two is what we should be looking to implement, but that should be done in a single PR rather than being split across two PRs, especially since some of the changes aren't compatible with one another. Because of that, I think it's better to close these out rather than letting them sit.

I've planned to combine these PRs myself, but I haven't gotten around to sinking my teeth into them just yet. If anyone else starts working on them, please let me know and I'll send over anything I have if I've started but don't think I'll be finishing soon.

**Pull request closed without merging**